pkg/ebpf/probes/probes.go

```diff
@@ -2,178 +2,503 @@ package probes
-		},
+		probes: allProbes,
 		module: module,
 	}, nil
 }
 
-//Attach the probe corresponding to the probe's handle
-func (p *probes) Attach(probe Handle) error {
-	if p.probes[probe] == nil {
-		return fmt.Errorf("failed to attach probe: doesn't exist")
+func (p *probes) Attach(args ...interface{}) error {
+	return p.attachOrDetach(true, args...)
+}
+
+func (p *probes) Detach(args ...interface{}) error {
+	return p.attachOrDetach(false, args...)
+}
+
+func (p *probes) attachOrDetach(attach bool, args ...interface{}) error {
+	var handle EventHandle
+	var netIface *net.Interface
+
+	// variatic function argument resolution
+	for _, arg := range args {
+		switch a := arg.(type) {
+		case *net.Interface:
+			netIface = a
+		case EventHandle:
+			handle = a
+		}
+	}
+
+	if _, ok := p.probes[handle]; !ok {
+		return fmt.Errorf("incorrect event handle provided")
 	}
 
-	return p.probes[probe].attach(p.module)
+	switch b := p.probes[handle].(type) {
+	case *traceProbe:
+		if attach {
+			return b.attach(p.module)
+		}
+		return b.detach()
+	case *tcProbe:
+		if netIface == nil {
+			return fmt.Errorf("program %s needs a net interface", b.programName)
+		}
+		if attach {
+			return b.attach(p.module, netIface)
+		}
+		return b.detach()
+	}
+
+	return nil
 }
 
-//Detach the probe corresponding to the probe's handle
-func (p *probes) Detach(probe Handle) error {
-	if p.probes[probe] == nil {
-		return fmt.Errorf("failed to detach probe: doesn't exist")
+func (p *probes) DetachAll() error {
+	var err error
+
+	for _, in := range p.probes {
+		switch i := in.(type) {
+		case *tcProbe:
+			err = i.detach()
+			if err != nil {
+				return err
+			}
+		case *traceProbe:
+			err = i.detach()
+			if err != nil {
+				return err
+			}
+		}
 	}
 
-	return p.probes[probe].detach()
+	return err
 }
 
-// ProbeType is an enum that describes the mechanism used to attach the event
-// Kprobes are explained here: https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#1-kp
-// Tracepoints are explained here: https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#3-tracep
-// Raw tracepoints are explained here: https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#7-raw-tracep
+//
+// probe
+//
+
 type probeType uint8
 
 const (
-	kprobe = iota
-	kretprobe
-	tracepoint
-	rawTracepoint
+	kprobe        = iota // github.com/iovisor/bcc/blob/master/docs/reference_guide.md#1-kp
+	kretprobe            // github.com/iovisor/bcc/blob/master/docs/reference_guide.md#1-kp
+	tracepoint           // github.com/iovisor/bcc/blob/master/docs/reference_guide.md#3-tracep
+	rawTracepoint        // github.com/iovisor/bcc/blob/master/docs/reference_guide.md#7-raw-tracep
 )
 
-//internal attach function - takes *bpf.Module the from the owning probes struct as a parameter
-func (p *probe) attach(module *bpf.Module) error {
-	if p.link != nil {
-		return nil
+type Probe interface {
```

Comment:
So this interface isn't the one you actually used for `traceProbe` and `tcProbe`, which DO have the same function signatures, just not the ones in the interface here.
I suggest we change the interface to fit what's actually used so we can use it in the `Probes` map, which could also remove a lot of the type casts (i've marked one such place with a review comment for example) making the code more readable.

So it should be (in pseudocode):
```
attach(args...)
detach(args...)
autoload(bool)
```